File: leet-code/641. Design Circular Deque/DesignCircularDeque641.java
public class DesignCircularDeque641 {
    private final int maxSize;
    private int length = 0;
    private final DoublyLinkedList head;
    private final DoublyLinkedList tail;

    public DesignCircularDeque641(int k) {
        maxSize = k;
        head = new DoublyLinkedList(-1);
        tail = new DoublyLinkedList(-1);
        head.right = tail;
        tail.left = head;
    }

    public boolean insertFront(int value) {
        if (isFull()) {
            return false;
        }
        DoublyLinkedList node = new DoublyLinkedList(value);
        node.left = head;
        node.right = head.right;
        // node 는 left, right 로 방금 연결이 되었으나
        // head, head.right 는 node 와 아직 연결되지 않아서
        // 기존의 연결을 유지하고 있는 상태다
        // 여기서 node 와 연결하면서 head, head.right 의 연결이 바뀐다
        //
        // node 를 먼저 head, head.right 와 연결했기 때문에
        // head, head.right 의 연결 상태가 바뀌지 않아서
        // head.right 의 참조값을 유지하기 위해 별도의 변수로 선언하지 않아도 됐다
        head.right.left = node;
        head.right = node;
        length++;
        return true;
    }

    public boolean insertLast(int value) {
        if (isFull()) {
            return false;
        }
        DoublyLinkedList node = new DoublyLinkedList(value);
        node.right = tail;
        node.left = tail.left;
        tail.left.right = node;
        tail.left = node;
        length++;
        return true;
    }

    public boolean deleteFront() {
        if (isEmpty()) {
            return false;
        }
        DoublyLinkedList node = head.right;
        head.right = head.right.right;
        head.right.left = head;
        node.left = null;
        node.right = null;
        length--;
        return true;
    }

    public boolean deleteLast() {
        if (isEmpty()) {
            return false;
        }
        DoublyLinkedList node = tail.left;
        tail.left = tail.left.left;
        tail.left.right = tail;
        node.left = null;
        node.right = null;
        length--;
        return true;
    }

    public int getFront() {
        return isEmpty() ? -1 : head.right.value;
    }

    public int getRear() {
        return isEmpty() ? -1 : tail.left.value;
    }

    public boolean isEmpty() {
        return length == 0;
    }

    public boolean isFull() {
        return maxSize == length;
    }

    static class DoublyLinkedList {
        int value;
        DoublyLinkedList left;
        DoublyLinkedList right;

        DoublyLinkedList(int value) {
            this.value = value;
        }
    }
}

File: leet-code/641. Design Circular Deque/DesignCircularDeque641.kt
class DesignCircularDeque641_2(k: Int) {
    private val maxSize: Int = k;
    private var length: Int = 0;
    private var head: DoublyLinkedList = DoublyLinkedList(-1)
    private var tail: DoublyLinkedList = DoublyLinkedList(-1)

    init {
        head.right = tail;
        tail.left = head;
    }

    fun insertFront(value: Int): Boolean {
        if (isFull())
            return false
        val node: DoublyLinkedList = DoublyLinkedList(value)
        node.left = head
        node.right = head.right
        head.right!!.left = node
        head.right = node
        length += 1
        return true
    }

    fun insertLast(value: Int): Boolean {
        if (isFull())
            return false
        val node: DoublyLinkedList = DoublyLinkedList(value)
        node.left = tail.left
        node.right = tail
        tail.left!!.right = node
        tail.left = node
        length += 1
        return true
    }

    fun deleteFront(): Boolean {
        if (isEmpty())
            return false
        head.right = head.right!!.right
        head.right!!.left = head
        length -= 1
        return true
    }

    fun deleteLast(): Boolean {
        if (isEmpty())
            return false
        tail.left = tail.left!!.left
        tail.left!!.right = tail
        length -= 1
        return true
    }

    fun getFront(): Int {
        return if (isEmpty()) -1 else head.right!!.value
    }

    fun getRear(): Int {
        return if (isEmpty()) -1 else tail.left!!.value
    }

    fun isEmpty(): Boolean {
        return length == 0
    }

    fun isFull(): Boolean {
        return maxSize == length
    }

    class DoublyLinkedList(val value: Int) {
        var left: DoublyLinkedList? = null;
        var right: DoublyLinkedList? = null;
    }
}
